velox/functions/sparksql/ToPrettyString.h

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/expression/CastExpr-inl.h"
+#include "velox/functions/Udf.h"
+#include "velox/type/Conversions.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace detail {
+static const StringView kNull = "NULL";
+}
+
+template <typename TExec>
+struct ToPrettyStringFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  template <typename A>
+  void initialize(
+      const std::vector<TypePtr>& inputTypes,
+      const core::QueryConfig& /*config*/,
+      A* /*a*/) {
+    inputType_ = inputTypes[0];
+  }
+
+  template <typename TInput>
+  void call(out_type<Varchar>& result, const TInput& input) {
+    if constexpr (std::is_same_v<TInput, StringView>) {
+      result.setNoCopy(input);
+      return;
+    }
+    if constexpr (std::is_same_v<TInput, int32_t>) {
+      if (inputType_->isDate()) {
+        auto output = DATE()->toString(input);
+        result.append(output);
+        return;
+      }
+    }
+    const auto castResult =
+        util::Converter<TypeKind::VARCHAR, void, util::SparkCastPolicy>::
+            tryCast(input);
+    if (castResult.hasError()) {
+      result.setNoCopy(detail::kNull);
+    } else {
+      result.copy_from(castResult.value());
+    }
+  }
+
+  template <typename TInput>
+  void callNullable(out_type<Varchar>& result, const TInput* a) {
+    if (a) {
+      call(result, *a);
+    } else {
+      result.setNoCopy(detail::kNull);
+    }
+  }
+
+ private:
+  TypePtr inputType_;
+};
+
+template <typename TExec>
+struct ToPrettyStringVarbinaryFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  template <typename TInput>
+  void callNullable(out_type<Varchar>& result, const TInput* input) {
+    if (input) {
+      result.resize(1 + 3 * input->size());
+      char* const startPosition = result.data();
+      char* pos = startPosition;
+      *pos++ = '[';
+      for (auto i = 0; i < input->size(); i++) {
+        auto formated = fmt::format("{:X}", input->data()[i]);
+        *pos++ = formated.data()[0];
+        *pos++ = formated.data()[1];
+        *pos++ = ' ';
+      }
+      *--pos = ']';
+    } else {
+      result.setNoCopy(detail::kNull);
+    }
+  }
+};
+
+template <typename TExec>
+struct ToPrettyStringTimeStampFunction {
```

Comment:
This will throw exception in 
```
template <>
struct ToPrettyStringFunction<Timestamp> {
  VELOX_DEFINE_FUNCTION_TYPES(Timestamp);

class "facebook::velox::Timestamp" has no member "resolver"
ToPrettyString.h(99, 3): detected during instantiation of type "facebook::velox::functions::sparksql::ToPrettyStringFunction<facebook::velox::Timestamp>::arg_type<facebook::velox::Varchar>" at line 99
```

Reply:
If don't define function types, out_type<Varchar> will throw exception